retrieval/scrape.go

```diff
@@ -763,6 +767,14 @@ loop:
 			var lset labels.Labels
 			mets := p.Metric(&lset)
 
+			for _, l := range lset {
+				if !utf8.ValidString(l.Value) {
+					err = errInvalidLabelValue
+					sl.l.With("timeseries", string(met)).Debug("Invalid UTF-8 label value")
+					break loop
+				}
+			}
```

Comment:
Same with pkg/textparse if you put the check into `Metric()`. I really think it belongs there. If I use a parsing package I expect the output to be valid. The usage invariant dictates that calling `Metric()` must only be skipped if the string has been seen before and been verified through `Metric()`.